File: services/java/institute/src/main/java/com/gradeloop/institute/model/Assignment.java
package com.gradeloop.institute.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import java.time.LocalDateTime;
import java.util.UUID;
import java.util.List;
import java.util.ArrayList;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Entity
@Table(name = "assignments")
public class Assignment {
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(nullable = false)
    private String name;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "course_id", nullable = false)
    @JsonIgnore
    private Course course;

    @JsonProperty("courseId")
    public UUID getCourseId() {
        return course != null ? course.getId() : null;
    }

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private AssignmentStatus status;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private AssignmentType type;

    private Integer autograderPoints;
    private Boolean allowManualGrading;

    private LocalDateTime releaseDate;
    private LocalDateTime dueDate;

    private Boolean allowLateSubmissions;
    private LocalDateTime lateDueDate;

    @OneToMany(mappedBy = "assignment", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private List<Question> questions = new ArrayList<>();

    @Embedded
    private GradingPolicy gradingPolicy;

    private Boolean enforceTimeLimit;
    private Integer timeLimit; // in minutes
    private Integer memoryLimit; // in MB

    private Boolean enableGroupSubmissions;
    private Integer groupSizeLimit;

    private Boolean enableLeaderboard;
    private Integer leaderboardEntries;

    public enum AssignmentStatus {
        DRAFT,
        PUBLISHED,
        ARCHIVED
    }

    public enum AssignmentType {
        LAB,
        EXAM,
        DEMO
    }
}

File: web/lib/validations/assignment.ts
import { z } from "zod";

export const assignmentTypeSchema = z.enum(["Lab", "Exam", "Demo"]);

export const createAssignmentSchema = z.object({
    type: assignmentTypeSchema,
    name: z.string().min(1, "Assignment name is required"),
    autograderPoints: z.number().min(0).optional().default(100),
    allowManualGrading: z.boolean().optional().default(false),
    releaseDate: z.date({
        message: "Release date is required",
    }),
    dueDate: z.date({
        message: "Due date is required",
    }),
    allowLateSubmissions: z.boolean().optional().default(false),
    lateDueDate: z.date().optional(),
    enforceTimeLimit: z.boolean().optional().default(false),
    timeLimit: z.number().min(1).optional(),
    enableGroupSubmissions: z.boolean().optional().default(false),
    groupSizeLimit: z.number().min(1).optional(),
    enableLeaderboard: z.boolean().optional().default(false),
    leaderboardEntries: z.number().min(1).optional(),
});

export type CreateAssignmentValues = z.infer<typeof createAssignmentSchema>;
